icu4c/source/i18n/measunit_extra.cpp

```diff
@@ -383,37 +401,57 @@ class Parser {
         return Token(match);
     }
 
+    /**
+     * Returns the next "single unit" via result.
+     *
+     * If a "-per-" was parsed, the result will have appropriate negative
+     * dimensionality.
+     *
+     * @param result Will be overwritten by the result, if status shows success.
+     * @param sawAnd If an "-and-" was parsed prior to finding the "single
+     * unit", sawAnd is set to true. If not, it is left as is.
+     * @param status ICU error code.
+     */
     void nextSingleUnit(SingleUnitImpl& result, bool& sawAnd, UErrorCode& status) {
-        sawAnd = false;
         if (U_FAILURE(status)) {
             return;
         }
 
-        if (!hasNext()) {
-            // probably "one"
-            return;
-        }
-
         // state:
         // 0 = no tokens seen yet (will accept power, SI prefix, or simple unit)
         // 1 = power token seen (will not accept another power token)
         // 2 = SI prefix token seen (will not accept a power or SI prefix token)
         int32_t state = 0;
         int32_t previ = fIndex;
 
-        // Maybe read a compound part
-        if (fIndex != 0) {
+        // Read a unit
+        while (hasNext()) {
```

Comment:
Can you make sure that your code doesn't allow identifiers such as `square-per-meter` or `metersecond` to sneak through?

The logic before was basically, if we aren't at the beginning of the string, then a TYPE_COMPOUND_PART token is *required* here.  Not optional, but required.

I think your new logic should be, if we are at the beginning of the string, then allow an optional 'per-'; otherwise, require a '-', '-per-', or '-and-'.  Maybe that requires putting `per-` in its own section in the trie; that's probably fine.

Reply:
Agreed - also noticed this just after I had "sent it for review" - will deal with it when I'm fresh in the morning, demonstrating the failures with proper unit tests before fixing it.